pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -178,11 +183,13 @@ func removeResources(elements []common.ClusterObjectReference, index int) []comm
 // addServiceStatus recorde the whole component status.
 // reconcile run at single threaded. So there is no need to consider to use locker.
 func (h *AppHandler) addServiceStatus(cover bool, svcs ...common.ApplicationComponentStatus) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
 	for _, svc := range svcs {
 		found := false
 		for i := range h.services {
 			current := h.services[i]
-			if current.Name == svc.Name && current.Env == svc.Env {
+			if current.Name == svc.Name && current.Env == svc.Env && current.Namespace == svc.Namespace && current.Cluster == svc.Cluster {
```

Comment:
will it possible that the namespace and cluster is empty?

Reply:
If empty, the recorded will also be empty. If not equal, a new service will be added.